com.unity.netcode.gameobjects/Runtime/Serialization/FastBufferReader.cs

```diff
@@ -0,0 +1,987 @@
+                {
+                    ReadBytes((byte*)native, target * sizeof(char));
+                }
+            }
+        }
+
+        /// <summary>
+        /// Reads a string.
+        /// NOTE: ALLOCATES
+        ///
+        /// "Safe" version - automatically performs bounds checking. Less efficient than bounds checking
+        /// for multiple reads at once by calling VerifyCanRead.
+        /// </summary>
+        /// <param name="s">Stores the read string</param>
+        /// <param name="oneByteChars">Whether or not to use one byte per character. This will only allow ASCII</param>
+        public unsafe void ReadValueSafe(out string s, bool oneByteChars = false)
+        {
+#if DEVELOPMENT_BUILD || UNITY_EDITOR
+            if (m_InBitwiseContext)
+            {
+                throw new InvalidOperationException(
+                    "Cannot use BufferReader in bytewise mode while in a bitwise context.");
+            }
+#endif
+
+            if (!VerifyCanReadInternal(sizeof(uint)))
+            {
+                throw new OverflowException("Reading past the end of the buffer");
+            }
+
+            ReadValue(out uint length);
+
+            if (!VerifyCanReadInternal((int)length * (oneByteChars ? 1 : sizeof(char))))
+            {
+                throw new OverflowException("Reading past the end of the buffer");
+            }
+            s = "".PadRight((int)length);
+            int target = s.Length;
+            fixed (char* native = s)
+            {
+                if (oneByteChars)
+                {
+                    for (int i = 0; i < target; ++i)
+                    {
+                        ReadByte(out byte b);
+                        native[i] = (char)b;
+                    }
+                }
+                else
+                {
+                    ReadBytes((byte*)native, target * sizeof(char));
+                }
+            }
+        }
+
+        /// <summary>
+        /// Writes an unmanaged array
+        /// NOTE: ALLOCATES
+        /// </summary>
+        /// <param name="array">Stores the read array</param>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public unsafe void ReadValue<T>(out T[] array) where T : unmanaged
+        {
+            ReadValue(out int sizeInTs);
+            int sizeInBytes = sizeInTs * sizeof(T);
+            array = new T[sizeInTs];
+            fixed (T* native = array)
+            {
+                byte* bytes = (byte*)(native);
+                ReadBytes(bytes, sizeInBytes);
+            }
+        }
+
+        /// <summary>
+        /// Reads an unmanaged array
+        /// NOTE: ALLOCATES
+        ///
+        /// "Safe" version - automatically performs bounds checking. Less efficient than bounds checking
+        /// for multiple reads at once by calling VerifyCanRead.
+        /// </summary>
+        /// <param name="array">Stores the read array</param>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
```

Comment:
Just a note, the JITter won't inline this because the function contains throw statements.